docs/proposals/image-updater-first-class-support.md

```diff
@@ -0,0 +1,239 @@
+It also doesn't store any state, and is, as such, unable to provide some key features that are discussed in further sections. Consolidating this configuration into the Application CR as first class fields in the spec and status is a natural next step in the Image Updater's journey. 
+
+### Goals
+
+- Propose updated Application CR fields to include configuration scoped for image updates
+- Make a case to merge Image Updater controller code into core Argo CD as its own controller 
+- Outline future enhancements planned for Image Updater and its growth within Argo CD 
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+### Use cases
+
+#### Use case 1:
+As a User of Argo CD, I would like to be able to have my application images updated automatically within expressed constraints when there is a new version available, out of the box with Argo CD. 
+
+#### Use case 2:
+As a user of Argo CD and Argo CD Image Updater, I would like to be able to express advanced Image update configurations through custom resource fields in my application.
+
+For users wanting to use Image Updater in production, it is common to have a strict set of restrictions in place in terms of which images can/should be automatically updated, where these images should come from, which reposoitry/branch the commits should be made to and what credentials should be used among many others. It is important to have a set of first class custom resource fields to be able to express this information accurately and reliably. This would add considerable value to the user experience, including but not limited to increased ease of use, stronger validation and standardization for these fields.
+ 
+#### Use case 3:
+As a user of Argo CD and Argo CD Image Updater, I would like to have image-updater honor rollbacks to my applications by leveraging status information stored in my application.
+
+### Implementation Details/Notes/Constraints [optional]
+
+
+- Argo CD Image Updater code base will need to be updated to look for application resources in namespaces other than the control-plane namespace when configured to use the k8s api instead of the argocd api to retrieve resources (see https://argocd-image-updater.readthedocs.io/en/stable/install/installation/#installation-methods). Currently it only recognizes applications present in the control plane namespace when installed in this manner. Exposed prometheus metric labels will need to be updated to take application namespace into account for accurate filtering. 
+
+- Image Updater controller code will need to be updated to watch and react to changes in newly defined application CR fields, and code to parse and react to annotations will need to be removed. 
+
+- Image Updater git write back code will need to be modified to figure out the correct set of manifest updates to be written back to each source repository in case the application has multiple source repositories. 
```

Comment:
>Instead of pushing changes back in git, which could lead to many different types of issues, why not relying on ignoreDifferences configurations, setting it dynamically removing the image field from the diff?

Would you mind elaborating on what specific issues you see with writing back to Git?

I can provide some additional reasoning and context for why I think we do need this:

- Writing back to Git would provide true persistence to configuration. It is also the most GitOps aligned method to keep track of changes
- In terms of disaster recovery, if the Argo CD instance ever goes down, having the updated state of the application written back to git would allow us to restore it to the cluster, along with the correct status information
- Having changes written back to Git allows replication of state easily in case of needing to ramp up new clusters with the same set of updated applications, for example.
- Git write back is currently an important feature for a lot of users of Image Updater, and is the preferred option over the existing pseudo-persistent alternative. If we were to get rid of this feature it would affect a lot of users that might be expecting to see this feature make it to Argo CD, and they might consider not using Image Updater at all

Some additional context on how git write back would work in Image Updater:
- Image Updater does not do any of the heavy lifting of git handling like repo-server does. The changes it makes are pretty minimal. It only creates a new file called .argocd-source-<appName>.yaml that contains the helm/kustomize fields for the updated information and adds it to the repo. We wouldn’t need to worry about direct changes to manifests or merge conflicts etc as such 
- Image Updater can use a dedicated secret for git write back credentials and does not have to access credentials stored in argo-cd (separation of secrets may provide better security)
- When Image Updater commits (or raises a PR) against a branch that is not tracked by the application, it is the user’s responsibility to make sure the changes are propagated to the branch tracked by the application to see the changes on the cluster. Image Updater will be able to use stored state to discern that repeat commits/PRs are not needed for the same update (if changes have not made it to the cluster yet)

Reply:
git writeback is a big draw for image updater users. It provides a full audit trail. I think it makes sense to try to keep it.